### PR TITLE
Pull request to merge Rui's work on path visualization and exploration

### PR DESCRIPTION
**Pull request by cyrille-artho**

### CODE REVIEW

**Comment by cyrille-artho** [comment]
Looks good, except for a few minor points and the "!" in "size".
Furthermore, please write a paragraph in README and CHANGELOG to explain what the new features are, and that these features have been added. It would also help to add a brief wiki page with an overview. You can describe things in more detail later in the wiki (perhaps wikifying the text from the upcoming paper).

**Comment by wruiwr** [comment]
@cyrille-artho  I wonder what the problem with "!" in "size".
I think the size attribute can scale the graphs down (by specifying the maximum bounding box of drawing in inches.), since when generating pdfs I got error, such as Error: canvas size (34642,3608) exceeds PDF limit (14400).
According to the document, If size ends in an exclamation point (!), then it is taken to be the desired size. 
Sure that if the graph is small already, then we do not need to use "!". I can remove it if you want.
what are other minor issues?

**Comment by wruiwr** [comment]
@cyrille-artho Yeah sure, the new features about smart search will be added later into README/WIKI.

File: log/nio-path-point-100.dot [hunk 1]
```diff
@@ -1,6 +1,6 @@
 digraph model {
   orientation = portrait;
-  graph [ rankdir = "TB", ranksep="0.1", nodesep="0.05" ];
+  graph [ rankdir = "TB", ranksep="0.1", nodesep="0.05", size="8!" ];
```
[the remark this answers is not in the document]
Comment (wruiwr) [reply log/nio-path-point-100.dot]: Explained in new version

File: src/main/scala/modbat/mbt/Modbat.scala [hunk 2]
```diff
@@ -542,11 +547,125 @@ object Modbat {
...
+      case "heur"   => heuristicChoice(choices, totalW)
+    }
+  }
+
+  def heuristicChoice(choices: List[(MBT, Transition)],
+                      totalW: Double): (MBT, Transition) = {
+    // compute choice based on bandit UCB  - Rui
+    val choice = banditUCBChoice(choices, totalW)
+    choice
+    //System.exit(1)
```
[the remark this answers is not in the document]
Comment (wruiwr) [reply src/main/scala/modbat/mbt/Modbat.scala]: removed

File: src/main/scala/modbat/mbt/Modbat.scala [hunk 3]
```diff
@@ -542,11 +547,125 @@ object Modbat {
...
+    Log.debug("failed assertion reward:" + Main.config.assertFailReward)
+
+    val currentStateCount = choices.head._1.currentState.coverage.count
+    val transCountLst = choices.map(_._2.coverage.count)
+//    val precondFailedCountLst = choices.map(_._2.coverage.precond.countPrecondFailed)
+    val precondFailedCountLst =
+      choices.map(_._2.coverage.expectedReward.countPrecondFail)
+
+//    Log.debug(
+//      "### list of failed assertion counts:" + choices.map(
```
[the remark this answers is not in the document]
Comment (wruiwr) [reply src/main/scala/modbat/mbt/Modbat.scala]: Keeping this in code with "Log,debug"

File: src/main/scala/modbat/mbt/Modbat.scala [hunk 4]
```diff
@@ -542,11 +547,125 @@ object Modbat {
...
+//        _._2.coverage.assertCount.countAssertFailed))
+//    Log.debug(
+//      "### list of passed assertion counts:" + choices.map(
+//        _._2.coverage.assertCount.countAssertPassed))
+//
+//    Log.debug(
+//      "*** list of passed precondition counts:" + choices.map(
+//        _._2.coverage.precond.countPrecondPassed))
+
+//    todo: the debug code:
```
[the remark this answers is not in the document]
Comment (wruiwr) [reply src/main/scala/modbat/mbt/Modbat.scala]: removed "todo" comments

File: src/main/scala/modbat/mbt/Modbat.scala [hunk 5]
```diff
@@ -686,32 +805,39 @@ object Modbat {
...
+    : ((TransitionResult, RecordedTransition), PathResult) = {
     var successors = allSuccessors(null)
     var allSucc = successors
     var totalW = totalWeight(successors)
     var backtracked = false // boolean var for backtracked case -Rui
     while (!successors.isEmpty && (totalW > 0 || !MBT.transitionQueue.isEmpty)) {
       val localStoredRNGState = MBT.rng.asInstanceOf[CloneableRandom].clone
-      if (MBT.rng.nextFloat(false) < Main.config.abortProbability) {
+      val abortProbability = MBT.rng.nextFloat(false)
+      if (/*MBT.rng.nextFloat(false)*/ abortProbability < Main.config.abortProbability) {
```
[the remark this answers is not in the document]
Comment (wruiwr) [reply src/main/scala/modbat/mbt/Modbat.scala]: removed

**Comment by cyrille-artho** [comment]
Looks good!
Please write a few test cases (in `bin/test.sh`). Use very small examples for this with a fixed random seed, to generate a simple output on coverage (and the number of paths etc.) and some variants that also generate the .dot file. Try to use a debug level that generates as little output as necessary to see some detail. (As long as we get an indication of the overall results, there is no need for finer-grained debug output, which tends to be brittle to changes.)

File: bin/test.sh [hunk 6]
```diff
@@ -233,14 +229,62 @@ savemv modbat.examples.NioSocket1-root-StateGraph.dot nio-path-state-500.dot
...
 
+## heuristic search 10 test cases
+run 0 $APP -n=10 -s=1 --no-redirect-out --log-level=debug --dotify-path-coverage --search=heur \
+	modbat.examples.NioSocket1
+savemv modbat.examples.NioSocket1-root-StateGraph.dot nio-path-state-heur-10.dot
+checkfile nio-path-state-heur-10.dot
+savemv modbat.examples.NioSocket1-root-PointGraph.dot nio-path-point-heur-10.dot
+checkfile nio-path-point-heur-10.dot
+## heuristic search 20 test cases
+run 0 $APP -n=20 -s=1 --no-redirect-out --log-level=debug --dotify-path-coverage --search=heur \
```
Comment (cyrille-artho) [inline bin/test.sh]: Thank you for adding so many tests.
Is there something that changes in particular in the output for n=20, 50, 150, 200?
If not, we can probably remove some of these tests. It is easier to maintain two or three variants than so many.

File: src/main/scala/modbat/mbt/Main.scala [hunk 7]
```diff
@@ -5,7 +5,7 @@ import modbat.config.Version
 import modbat.log.Log
 
 object Main {
-  val config = new Configuration()
+  var config = new Configuration()
```
Comment (cyrille-artho) [inline src/main/scala/modbat/mbt/Main.scala]: Is this change from `val` to `var` needed?

Comment (wruiwr) [reply src/main/scala/modbat/mbt/Main.scala]: Oops...I'm not sure how it is changed. Maybe a mistake happened when editing, but I will change it back.

**Pull request merged**